core/modulesource.go

```diff
@@ -261,6 +265,153 @@ func (src *ModuleSource) AutomaticGitignore(ctx context.Context) (*bool, error)
 	return modCfg.Codegen.AutomaticGitignore, nil
 }
 
+// LoadContext loads a directory from the module context directory.
+//
+// If the module is local, it will load the directory from the local source
+// directly from the host.
+//
+// If the module is git, it will load the directory from the git repository
+// using its context directory.
+//
+// The path is relative to the module context directory, even if it's absolute.
+//
+// Improvement for later: use `.gitignore` to excludes paths from the context.
+func (src *ModuleSource) LoadContext(ctx context.Context, dag *dagql.Server, path string, ignore []string) (inst dagql.Instance[*Directory], err error) {
+	// We exclude .git by default because it's not useful and tends to invalidate the cache.
+	excludes := []string{"**/.git"}
+	includes := []string{}
+
+	for _, p := range ignore {
+		if strings.HasPrefix(p, "!") {
+			includes = append(includes, p[1:])
+		} else {
+			excludes = append(excludes, p)
+		}
+	}
+
+	// If path is not absolute, it's relative to the module root directory.
+	if !filepath.IsAbs(path) {
+		moduleRootPath, err := src.SourceRootSubpath()
+		if err != nil {
+			return inst, fmt.Errorf("failed to get source root subpath for relative context path: %w", err)
+		}
+
+		path = filepath.Join(moduleRootPath, path)
+	}
+
+	switch src.Kind {
+	case ModuleSourceKindLocal:
+		bk, err := src.Query.Buildkit(ctx)
+		if err != nil {
+			return inst, fmt.Errorf("failed to get buildkit api: %w", err)
+		}
+
+		slog.Debug("moduleSource.LoadContext: loading contextual directory from local source", "path", path, "kind", src.Kind)
+
+		ctxPath, _, err := src.ResolveContextPathFromCaller(ctx)
+		if err != nil {
+			return inst, fmt.Errorf("failed to resolve context path: %w", err)
+		}
```

Comment:
There's a *very* subtle issue with using this in this way. If we have a `Directory` and we call `AsModule` on it, then the path defined by this module has *no corresponding context directory on the host*, so we need to instead get it from the `src.ContextDirectory()`.

This is a real concern for the builtin sdk runtimes (which use this). However, there's some also some *supreme* strangeness with how `ResolveContextPathFromCaller` works at all - specifically, it's a surprise it works at all. The following scenario on this PR breaks:

```
$ git init
$ mkdir ctxmod && cd ctxmod
$ dagger init --sdk=go --source=.
$ cat <<EOF > main.go
package main

import (
	"context"
)

type Ctxmod struct{}

func (t *Ctxmod) Dirs(
	ctx context.Context,
	// +defaultPath="/"
	root *Directory,
	// +defaultPath="."
	relativeRoot *Directory,
) ([]string, error) {
	res, err := root.Entries(ctx)
	if err != nil {
		return nil, err
	}
	relativeRes, err := relativeRoot.Entries(ctx)
	if err != nil {
		return nil, err
	}
	return append(res, relativeRes...), nil
}
EOF
$ dagger call -m . dirs
...
13  : Ctxmod.dirs: [String!]!
13  : Ctxmod.dirs ERROR [0.0s]
13  : ! failed to load contextual arg "root": failed to load contextual directory "/": failed to resolve context path: failed to stat source root "./ctxmod": failed to stat path: failed to receive file bytes message: rpc error: code = NotFound desc = rpc error: code = NotFound desc = stat path: lstat /home/jedevc/Documents/Projects/dagger-test-modules/ctxmod/ctxmod: no such file or directory
```

Effectively, we're applying the subpath from the root *twice*. This appears to be an issue from *before* this PR.

I think ResolveContextPathFromCaller's implementation is actually wrong:
- it first gets the root subpath (relative to the context)
- it then tries to look to find that root subpath in the current directory on the client (using StatCallerHostPath, which is relative to the current directory)
- but the current directory on the client is not neccessarily the subpath - the client ends up looking for <abspath-to-context>/subpath/subpath somehow

Again, this is a real issue, and needs solving before we can merge this PR. I'm happy to pick these up, *but* I think these take longer than we have before the v0.12 release - so I'm going to push these into the v0.12.1 milestone - in the interest of shipping *something* this week.

---

A few TODOs:
- A `LocalSource` should track *where* the context directory it's been loaded from is. This removes the complexity of this specific dance.
- We should create a `ScratchSource` type that can be created for `Directory.AsModule` - this type *doesn't* map to a directory on the host, so we should avoid all chances of confusion, which this PR steps into.

Some other notes from investigating:
- Why are there two implementations of `findUp`? One on the cli, one on the engine. This is *super* confusing, and isn't helping simplify file interactions.
- This extends into a "who is responsible for loading a module" problem - it's really unclear how the cli/server interact to load a module, with lots of jumping between cli/filesync/engine
	- @helderco has also noticed subtle bugs around this:
		> - One example is when using an "unbundled" sdk with --sdk and --source. The path to the SDK is calculated in relation to --source, which is wrong
		> - There's another issue where if a module fails to load, it reverts to a parent module (parent as in the filesystem tree)

Reply:
> Why are there two implementations of findUp? One on the cli, one on the engine. This is super confusing, and isn't helping simplify file interactions.
This extends into a "who is responsible for loading a module" problem - it's really unclear how the cli/server interact to load a module, with lots of jumping between cli/filesync/engine

The duplication between CLI+engine used to be 100x worse (essentially two parallel implementations of the entire process). 
* e.g. git sourced modules can have "local deps" (which are "local" to the git repo) and it used to be that there were completely separate implementations of loading those "local" deps and loading actually local (on the caller's disk) modules even though the process is 95% the same. Many other similar situations of duplication used to exist.

While we were getting ready for the modules release there were too many bugs that when fixed just spawned more bugs, so I started centralizing the logic in the server-side to help at least keep the complication in mostly one place, which helped at least put out the fires in our hair at that time.

The current state of things where there's still some remaining logic in the CLI is just because on a practical level I had hit the timebox on this whole area and the known fires were out. So no _good_ reason there's duplication remaining, just pragmatic ones from the past.

Point being, yes we should centralize the loading logic even more and get rid of the remaining split. Because the server-side *has* to be the place for a large subset of this logic (e.g. git sourced modules and their "local" deps), it's the only reasonable place to centralize it, so I think we should continue going down that path.

I'll just warn that even though I always imagined all this to be complicated going into it, it still always ended up being several times more complicated than I imagined 😄, so I would just expect this to take a bit to get working in every case. We have a shit-ton of integ tests for all this path/loading stuff, but I'm sure it's still missing quite a bit, so while fixing what remains here we likely need to upgrade to a _metric_ shit-ton of integ tests.

> I think ResolveContextPathFromCaller's implementation is actually wrong:
> 
> * it first gets the root subpath (relative to the context)
> * it then tries to look to find that root subpath in the current directory on the client (using StatCallerHostPath, which is relative to the current directory)
> * but the current directory on the client is not neccessarily the subpath - the client ends up looking for /subpath/subpath somehow

Yeah it does appear to be inconsistent whether `SourceRootSubpath` is made to be relative to the cwd or not
* We do it for abs paths here https://github.com/sipsma/dagger/blob/6d2bd214bcf6c722399ef820c025193eafd9da0e/core/schema/modulesource.go#L49-L49
* We do it during `dagger init` since it takes an explicit path to the new source: https://github.com/sipsma/dagger/blob/6d2bd214bcf6c722399ef820c025193eafd9da0e/cmd/dagger/module.go#L112-L112
* But don't appear to for user-supplied relative paths when just using the `-m` flag https://github.com/sipsma/dagger/blob/6d2bd214bcf6c722399ef820c025193eafd9da0e/cmd/dagger/module.go#L575-L575

So it seems like we could get away with just fixing that last point. _However_, this is a very good example of the remaining split between CLI+engine being confusing and requiring jumping all over the place. Centralizing more will help a lot.

---

I honestly think at this point we should just go all the way and add core APIs that do literally everything so that each relevant CLI subcommand (dagger init/install/develop/call) is just a one-line call to the core API.
* This aligns with the discussion in https://github.com/dagger/dagger/issues/6947#issuecomment-2135899729 around it making sense to turn the whole CLI into a stub for just calling core functions